packer/core.go

```diff
@@ -404,44 +432,84 @@ func (c *Core) init() error {
 
 	for i := 0; i < 100; i++ {
 		shouldRetry = false
+		changed = false
+		deleteKeys := []string{}
 		// First, loop over the variables in the template
-		for k, v := range allVariables {
+		for _, kv := range sortedMap {
 			// Interpolate the default
-			renderedV, err := interpolate.RenderRegex(v, ctx, renderFilter)
+			renderedV, err := interpolate.RenderRegex(kv.Value, ctx, renderFilter)
 			switch err.(type) {
 			case nil:
 				// We only get here if interpolation has succeeded, so something is
 				// different in this loop than in the last one.
 				changed = true
-				c.variables[k] = renderedV
+				c.variables[kv.Key] = renderedV
 				ctx.UserVariables = c.variables
+				// Remove fully-interpolated variables from the map, and flag
+				// variables that still need interpolating for a repeat.
+				done, err := isDoneInterpolating(kv.Value)
+				if err != nil {
+					return ctx, err
+				}
+				if done {
+					deleteKeys = append(deleteKeys, kv.Key)
```

Comment:
if you range over the keys slice you could just delete the key from it here, then the loop for deleting the key-value from the map wouldn't be necessary.

Reply:
I tried implementing the slice ranging like you did for HCL but after 3 hours and still having a ton of broken tests I opted for a less major change. I'm always very uncomfortable adding/removing items from the same thing I'm looping over, especially when it has to do with indexing. I think it makes the code harder to conceptualize.